src/server/scripts/Kalimdor/Firelands/boss_ragnaros_firelands.cpp

```diff
@@ -0,0 +1,5525 @@
+        }
+};
+
+// http://www.wowhead.com/npc=54127/dreadflame
+class npc_fl_ragnaros_dreadflame : public CreatureScript
+{
+    public:
+        npc_fl_ragnaros_dreadflame() : CreatureScript("npc_fl_ragnaros_dreadflame") {}
+
+        struct npc_fl_ragnaros_dreadflameAI : public PassiveAI
+        {
+            npc_fl_ragnaros_dreadflameAI(Creature* creature) : PassiveAI(creature) {}
+
+            void EnterEvadeMode(EvadeReason /*reason*/) override { }
+
+            void Reset() override
+            {
+                Dreadflames.clear();
+                for (const auto ConstDreadflame : ConstDreadflames)
+                {
+                    DreadflameInfo* dreadflame = new DreadflameInfo();
+                    dreadflame->Position = ConstDreadflame;
+                    dreadflame->Owner = ObjectGuid::Empty;
+                    Dreadflames.push_back(dreadflame);
+                }
+
+                for (auto sourceDreadflame : Dreadflames)
+                    for (auto targetDreadflame : Dreadflames)
+                        if (sourceDreadflame != targetDreadflame && sourceDreadflame->Position.IsInDist2d(targetDreadflame->Position.GetPositionX(), targetDreadflame->Position.GetPositionY(), 5.1f))
+                            sourceDreadflame->NearbyDreadflames.push_back(targetDreadflame);
+
+                me->CastSpell(me, SPELL_DREADFLAME_SPREAD_CONTROL_AURA, true);
+                me->CastSpell(me, SPELL_DREADFLAME_DELUGE_CONTROL_AURA, true);
+                me->CastSpell(me, SPELL_DREADFLAME_SPAWN_CONTROL_AURA, true);
+                me->CastSpell(me, SPELL_DREADFLAME_DAMAGE_CONTROL_AURA, true);
+            }
+
+            void JustRegisteredDynObject(DynamicObject* dynObj) override
+            {
+                RegisterDreadflame(dynObj);
+            }
+
+            void JustUnregisteredDynObject(DynamicObject* dynObj) override
+            {
+                UnregisterDreadflame(dynObj);
+            }
+
+            void RegisterDreadflame(DynamicObject* dynObj)
+            {
+                for (uint32 i = 0; i < DreadFlamesMax; i++)
+                {
+                    if (std::sqrt(dynObj->GetExactDist2dSq(Dreadflames[i]->Position.GetPositionX(), Dreadflames[i]->Position.GetPositionY()) < 3.0f))
+                    {
+                        Dreadflames[i]->Owner = dynObj->GetGUID();
+                        break;
+                    }
+                }
+            }
+
+            void UnregisterDreadflame(DynamicObject* dynObj)
+            {
+                for (uint32 i = 0; i < DreadFlamesMax; i++)
+                {
+                    if (Dreadflames[i]->Owner == dynObj->GetGUID())
+                    {
+                        Dreadflames[i]->Owner = ObjectGuid::Empty;
+                        break;
+                    }
+                }
+            }
+
+            std::list<Position> GetDreadflameSpreadPositions() const
```

Comment:
Probably the method that has the most performance impact that needs to be adjusted right now.
Maybe instead of recalculating dreadflame positions in proximity on each tick (every 3.5 seconds), this could be cached and updated every time a dreadflame is added/removed instead.